Add elixir support via dogma

kthelgason opened the pull request:
This PR adds support for the Elixir language (as discussed in #877).
This relies on [dogma](https://github.com/lpil/dogma) for syntax checking, and so does not execute the code being checked as `elixirc` would.

Edit (@lunaryorn): Connects to #877 

TODO:

- [x] ~~Add predicate to make sure checker is not loaded when `mix dogma` is not available.~~
- [x] Parse dogma's JSON output format.
- [x] Tests for JSON parsing

lunaryorn commented on flycheck.el:
```diff
@@ -6182,6 +6183,19 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+          (message) line-end))
+  :modes elixir-mode)
+
+
```
Please remove this blank line :)

kthelgason commented:
@lunaryorn removed blank line and force-pushed :) thanks for taking a look.

lunaryorn commented on flycheck.el:
```diff
@@ -6182,6 +6183,19 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+  "An Elixir syntax checker using the Dogma analysis tool.
+
+See URL `https://github.com/lpil/dogma/'."
+  :command ("mix" "dogma" "--format=flycheck" source-inplace)
```
`--format=flycheck`?! :blush:

`dogma` seems to [support JSON](https://github.com/lpil/dogma/blob/master/docs/reporters.md) output, though, and we generally prefer to parse structured output formats, which tend to be less fragile than plain text.

Would you mind to update this pull request to use an `:error-parser` that parses the JSON output of `dogma`?

lunaryorn replied:
By the way how does `mix` know that `dogma` is available?  Can we somehow test that to make sure that we don't enable this syntax checker when `mix` is there but `dogma` is not?

lunaryorn replied:
And last but not least, do we need `source-inplace`, or can we feed the buffer contents to `dogma` via standard input?

lunaryorn commented:
@kthelgason Thank you very much!  It's great news to have Elixir back in, thanks for your work!

I've left a couple of comments on the diff, would you take a look?

kthelgason commented:
Thanks for the review @lunaryorn! As for your comments:

> Would you mind to update this pull request to use an :error-parser that parses the JSON output of dogma?

I can do this, but it will involve more code and be more complicated. Here is an example of the JSON output: 
```js
{
  "metadata": {
    "dogma_version": "0.3.0",
    "elixir_version": "1.0.5",
    "erlang_version": "Erlang/OTP 10 [erts-7.0.3] [64-bit]",
    "system_architecture": "x86_64-apple-darwin14.5.0"
  },
  "files": [{
      "path": "lib/foo.ex",
      "errors": []
   }, {
      "path": "lib/bar.ex",
      "errors": [{
          "line": 1,
          "rule": "ModuleDoc",
          "message": "Module without @moduledoc detected"
       }, {
          "line": 14,
          "rule": "ComparisonToBoolean",
          "message": "Comparison to a boolean is useless"
       }
      ]
  }],
  "summary": {
    "error_count": 2,
    "inspected_file_count": 2
  }
}
```

It's up to you wether you think it's worth it :)

> By the way how does mix know that dogma is available? Can we somehow test that to make sure that we don't enable this syntax checker when mix is there but dogma is not?

This is an excellent point, and I don't really have a good answer, perhaps you have some ideas as to how to accomplish this?

> And last but not least, do we need source-inplace, or can we feed the buffer contents to dogma via standard input?

Unfortunately `dogma` does not currently support reading the source from stdin, but there is an open PR that implements that functionality. I can keep an eye on it and submit another PR when it lands.

lunaryorn commented:
@kthelgason 

> It's up to you wether you think it's worth it :)

Yes, I do.  Take a look at the `tslint` checker to see an example for JSON parsing.  And—if you've a little more time to spare—maybe also take a look at `test/specs/languages/test-typescript.el` to see a test case for JSON parsing, and probably add a test case for parsing dogma errors? :blush:

> This is an excellent point, and I don't really have a good answer, perhaps you have some ideas as to how to accomplish this?

Flycheck checkers can supply a `:predicate` to add additional checks.  Take a look at the `go-vet` syntax checker when looks at the installed Go tools to find out whether `vet` is available.  Can we do something similar for `mix`?  Is there a `mix commands` or similar which lists all available `mix` commands?

> Unfortunately dogma does not currently support reading the source from stdin, but there is an open PR that implements that functionality. I can keep an eye on it and submit another PR when it lands.

Would you mind to link this PR here so that we can track its process, too?

Meanwhile, would `source` instead of `source-inplace` work?  We generally prefer the former because it keeps temporary files out of the source tree…

kthelgason commented:
Certainly, I'll find some time to work on this in the coming days and ping you when I have something. Thanks for the suggestions!

> Flycheck checkers can supply a :predicate to add additional checks. Take a look at the go-vet syntax checker when looks at the installed Go tools to find out whether vet is available. Can we do something similar for mix? Is there a mix commands or similar which lists all available mix commands?

Yup, this would totally work. 

> Would you mind to link this PR here so that we can track its process, too?

https://github.com/lpil/dogma/pull/194

> Meanwhile, would source instead of source-inplace work? We generally prefer the former because it keeps temporary files out of the source tree…

Yes, that should definitely work as well. I'll change that.

lunaryorn commented:
@kthelgason Great, thank you very much :+1:

kthelgason commented:
Update:
I'm pretty much done rewriting the checker to use dogma's json format. I ran into some issues with the checker itself and I'm working to get them upstreamed. I'm still waiting to resolve some issues before I can fully test the code I've written here. Will update again when I've gotten this resolved.

kthelgason commented:
### Update:
There were some issues with using dogma as a mix command so I've added support for running dogma standalone. I'm going to submit a PR to dogma to get those changes in. Meanwhile I've updated this PR to work with that version of dogma. I've been running it in this configuration for a while and it seems to work well.

I apologise in advance if I'm doing something stupid in this PR, this is the first real code I write in elisp.

cc @lunaryorn

lunaryorn commented:
@kthelgason May I ask what's the status of this?  Is the change to dogma already submitted?  Is this PR ready for being merged?

kthelgason commented:
@lunaryorn Absolutely. The change has been pulled into dogma, so the code in this PR works with the current version. As for whether it's ready to be merged, you tell me 😄 from my perspective it's done.

lunaryorn commented:
@kthelgason Great, I'll review it again.

Would you like to join Flycheck to maintain our Elixir support?  You'd be taking a look at Elixir issues and pull request.

kthelgason commented:
I can absolutely do that. I intended to add support for another elixir syntax checker soon as well, and I'm totally willing to maintain this code.


		_____________________________
From: Sebastian Wiesner <notifications@github.com>
Sent: Sunday, July 17, 2016 12:41 pm
Subject: Re: [flycheck/flycheck] Add elixir support via dogma (#969)
To: flycheck/flycheck <flycheck@noreply.github.com>
Cc: Mention <mention@noreply.github.com>, Kári Tristan Helgason <kthelgason@gmail.com>




@kthelgason Great, I'll review it again.

Would you like to join Flycheck to maintain our Elixir support?  You'd be taking a look at Elixir issues and pull request.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

lunaryorn commented on test/specs/languages/test-elixir.el:
```diff
@@ -0,0 +1,77 @@
+;;; test-elixir.el ends here
+;;; (flycheck-parse-elixir-dogma json 'checker 'buffer)
+
+
```
Please remove excessive blank lines at the end of this file.

lunaryorn commented on test/specs/languages/test-elixir.el:
```diff
@@ -0,0 +1,77 @@
+                                                          :filename "lib/bar.ex")))))))
+
+;;; test-elixir.el ends here
+;;; (flycheck-parse-elixir-dogma json 'checker 'buffer)
```
What's the purpose of this?

lunaryorn commented on test/specs/languages/test-elixir.el:
```diff
@@ -0,0 +1,77 @@
+(require 'flycheck-buttercup)
+
+(describe "Language Elixir"
+          (describe "The Dogma error parser"
```
The indentation is wrong.

Please install and load buttercup, mark the whole buffer and run "indent-region".

lunaryorn commented on doc/languages.rst:
```diff
@@ -238,6 +238,12 @@ to view the docstring of the syntax checker.  Likewise, you may use
+
+   .. syntax-checker:: elixir-dogma
+
+      Check syntax and code style with `Dogma <https://github.com/lpil/dogma>`_
```
Please add a period at the end of this sentence.

lunaryorn commented:
Great, I'll add you to our org tomorrow.  

Meanwhile please address the remaining issues, rebase onto master and add a note about this change to CHANGES.rst.  Please also feel free to add yourself to our list of contributors in doc/community/people.rst. 😊

lunaryorn commented on flycheck.el:
```diff
@@ -6182,6 +6183,43 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+
+See URL `https://github.com/lpil/dogma' for more information
+about dogma."
+  (cl-flet ((alist-get (key alis) (cdr (assoc key alis))))
```
Could you probably write this function without this local alist-get function?  It's not a bad idea per se, not at all, but it's a style that we dont use in Flycheck, so for the sake of consistency I'd prefer if we could go without local functions here ☺️

lunaryorn replied:
Besides I think you can use assq instead of assoc here as all the keys are symbols.

kthelgason commented:
Thanks for the review @lunaryorn! I addressed your comments and amended the commit. I also added an entry to `CHANGES.rst` and to `people.rst`. Let me know if these are incorrectly formatted, and if you want me to add a language team for elixir as well 😃

lunaryorn commented on CHANGES.rst:
```diff
@@ -8,6 +8,10 @@
 
+- New syntax checkers:
+
+  - Elixir with ``dogma`` [GH-969]
```
I'm sorry to bother you again, but as a convention we list new syntax checkers first, as they are likely the most interesting change for users.

If you rebase onto master again, you'll see another entry for a new syntax checker.  Please add yours to it.

lunaryorn commented:
@kthelgason I've sent you an invitation to our org.  Accepting it will put you in a new Elixir team, and give you push access to Flycheck.  Please read through our [Contributor's Guide](http://www.flycheck.org/en/latest/contributor/contributing.html) for general information about contributing and through our [Maintainer's Guide](http://www.flycheck.org/en/latest/contributor/maintaining.html) for information specific to those with write access to our repository.

Please note that you won't be able to push to `master` directly as we require status checks for all contributors, but you'll be able to create your own branches, and you'll be able to merge approved pull requests.

kthelgason commented:
@lunaryorn Thanks, I've accepted the invite, and I'll make sure to read through the guides. The documentation for this project is really exemplary. I also rebased again and fixed the changelog!

Simplify commented:
@kthelgason Is there a reason why you choose this approach of using global installation of `dogma` instead of implementing your first idea: to add predicate into checker. You can check if `dogma` is installed by parsing output of `mix local`. Most developers will always install `dogma` locally using `mix.exs` file. If there are benefits for using global installation of `dogma` and this PR goes through, please explain in documentation that global installation of `dogma` is required.
You also mention that you are planning to add another syntax checker for elixir. I guess it's Credo. With this approach, users will be forced to install two packages globally (credo and bunt).

kthelgason commented:
@Simplify the reason is simply that dogma has to be installed globally, either as an erlang archive or as an escript executable (current approach). This is because in order for the `mix dogma` command to work it has to be invoked in the root dir of the project, and flycheck (AFAIK) invokes the executable in the parent dir of the file being checked.

Suggestions as to how to work around this are greatly appreciated, and I agree that this approach is limiting.

cpitclaudel commented:
@kthelgason We recently added support for `:working-directory`; could it be useful here?

Simplify commented:
Yes, you can find project directory by checking where `mix.exs` is located. There is always a possibility that project is so small that it doesn't uses `mix`. 
In this case It's possible  to use `flycheck-locate-config-file` to find where `mix.exs` is located, strip filename from result and feed it to `:working-directory`.

lunaryorn commented:
@Simplify We should rather use `locate-dominating-directory` directly for this purpose.

@kthelgason Would you take a look into this?  Or should we merge this PR now, and you'd investigate this possibility later?

Simplify commented:
@lunaryorn Yeah, `locate-domanating-directory` is best choice. `locate-config-file` was the first one that got into my mind. We can maybe do something similar as `projectile` in the future. Determining project directory is also useful for `node` projects is some cases, allows you to use packages from `node_modules/.bin`. Right now I have `.dir-locals.el` in every project that uses node in order to set exec path in GNU Emacs and use local packages per project.
In any case, documentation may need change in this PR: to explain that this checker needs global `dogma` installation or that `mix.exs` is required in every project if predicate and working-directory are implemented.

kthelgason commented:
@lunaryorn I'd like to look into this before merging this PR. There is another limitation that this would solve, namely that dogma doesn't find it's config file unless it's invoked from the root directory. I was planning on solving this in dogma, but it's even easier to fix it on the flycheck side. I have an 11 hour flight tomorrow morning, so that should give me a chance to look into this. If that doesn't work out the next two weeks are very busy for me (moving to another country) but I'll do what I can!

lunaryorn commented:
@kthelgason Please take all the time you need, I'll mark this PR as "blocked" meanwhile.  Please feel free to ask questions about if you have any.

I hope your move goes well!

kthelgason commented:
Added some commits implementing what we discussed.
There is one issue left: the predicate function does not respect `:working-directory`.
I'd like to fix this if someone can point me in the right direction.

Thanks for your patience with this, I've been quite busy with moving and starting a new job etc.

lunaryorn commented on flycheck.el:
```diff
@@ -6248,6 +6249,66 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+of the current file."
+  (or
+   (locate-dominating-file (buffer-file-name) "mix.exs")
+   (file-name-directory (buffer-file-name))))
```
Is there a reason why you didn't choose the standard `default-directory` as fallback here?

lunaryorn replied:
ping @kthelgason  :blush:

lunaryorn commented on flycheck.el:
```diff
@@ -6248,6 +6249,66 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+  (let ((mix (flycheck-checker-executable 'elixir-dogma))
+        (found))
+    (dolist (line (process-lines mix "help" ))
+      (if (string-match "dogma" line)
```
Use `string-match-p` instead of `string-match` here.  We don't need to modify the match data as we don't use it anywhere.

lunaryorn commented on flycheck.el:
```diff
@@ -6248,6 +6249,66 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+Used as a predicate for enabling the checker."
+  (let ((mix (flycheck-checker-executable 'elixir-dogma))
+        (found))
+    (dolist (line (process-lines mix "help" ))
```
I think I'd prefer a more functional approach.  What about `(seq-find (lambda (l) (string-match-p "dogma" l)) (process-lines …))`?

Besides, I think you should wrap `(ignore-errors)` around `(process-lines)`, just for the sake of safety.  `process-lines` signals an error if the command returns a non-zero exit code.

lunaryorn commented:
@kthelgason Does it matter that the predicate has a different default directory?  It doesn't seem to make us of it?  Or does `mix help` require a certain directory?

/cc @sergv As far as I remember you added support for `:working-directory`.  I'm not too familiar with your implementation anymore, so I'm not sure how hard it'd be to extend the scope of the working directory change to include the evaluation of the predicate.  Could you perhaps take a look?

kthelgason commented:
@lunaryorn Thanks for your excellent comments. I probably would have done things the way you suggest right away if I knew elisp a little better :smile: I'll get there.

As for the predicate having a default directory, `mix help` will only see the `dogma` command when invoked from the project root directory (on the same level as `mix.exs`) unfortunately, as this pretty much renders the checker useless.

lunaryorn commented:
@kthelgason Ok, so for the purpose of this PR I guess you can just `let`-bind `default-directory` to the result of the working directory function in the `:predicate`.  That's not as nice, but it gets the job done for this PR.

We can think about default directory for `:predicate` later.

sergv commented:
@lunaryorn It seems straightforward to make sure `:working-directory` is used for `:predicate`. I think it should be done in the following way: the `flycheck-define-generic-checker` function should wrap passed `:predicate` argument with code that changes directory to the result of `:working-directory`. This way all subsequent invocations of `:predicate` will automatically use correct directory. Would you agree with the approach?

lunaryorn commented:
@sergv It sounds reasonable for sure.  Would you like to implement this change?

sergv commented:
@lunaryorn Okay, I'll add it soon.

lunaryorn commented:
Blocked by #1040

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/flycheck/flycheck?pullRequest=969) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/flycheck/flycheck?pullRequest=969) before we can accept your contribution.<br/><hr/>**Kári Tristan Helgason** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/flycheck/flycheck?pullRequest=969) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/flycheck/flycheck?pullRequest=969) before we can accept your contribution.<br/><hr/>**Kári Tristan Helgason** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/flycheck/flycheck?pullRequest=969) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/flycheck/flycheck?pullRequest=969) before we can accept your contribution.<br/><hr/>**Kári Tristan Helgason** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/flycheck/flycheck?pullRequest=969) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/flycheck/flycheck?pullRequest=969) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: kthelgason<br/>:x: Kári Tristan Helgason<hr/>**Kári Tristan Helgason** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

lunaryorn commented:
Looks like you committed with a wrong mail address 😊

kthelgason commented:
I screwed up some settings and had to to battle with the CLA bot, luckily I won 😄 I cherry-picked @sergv's commit and tested it, I can confirm that the checker works brilliantly.

Also refactored according to the review, thanks for that @lunaryorn, great comments as usual!

lunaryorn commented:
@kthelgason Awesome, great news.  It's ready to be merged then.

But please remove @sergv's commit from this PR again, as we'll merge it separately n #1040 before merging this one.  Finally please squash all commits into a single one.

Sorry for being such a nuisance to you 😌

kthelgason commented:
@lunaryorn Done and done! Thanks for helping me out with this @lunaryorn and @sergv

cpitclaudel commented on flycheck.el:
```diff
@@ -6262,6 +6263,63 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+  "Come up with a suitable default directory to run CHECKER in.
+
+This will either be the directory that contains `mix.exs' or,
+if no such file is found in the directory hierarchy, ths directory
```
`ths → the`?

cpitclaudel commented on flycheck.el:
```diff
@@ -6262,6 +6263,63 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+(defun flycheck-elixir--parse-dogma-json (output checker buffer)
+  "Parse dogma errors from JSON OUTPUT.
+
+CHECKER and BUFFER denoted the CHECKER that returned OUTPUT and
```
`denoted → denote`?

cpitclaudel commented on flycheck.el:
```diff
@@ -6262,6 +6263,63 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+Check by parsing the output of `mix help'.
+Used as a predicate for enabling the checker."
+  (let ((mix (flycheck-checker-executable 'elixir-dogma)))
+    (seq-find (lambda (l) (string-match-p "dogma" l))
```
This lambda could be written with `apply-partially` (I don't know if it would actually be better)

kthelgason replied:
I feel that would only serve to make it less clear in this case.

cpitclaudel commented on flycheck.el:
```diff
@@ -6262,6 +6263,63 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
+   default-directory))
+
+(defun flycheck-elixir--parse-dogma-json (output checker buffer)
+  "Parse dogma errors from JSON OUTPUT.
```
`dogma → Dogma`?

cpitclaudel commented on flycheck.el:
```diff
@@ -6262,6 +6263,63 @@ Requires DMD 2.066 or newer.  See URL `http://dlang.org/'."
   :modes d-mode)
 
+(defun flycheck-elixir--find-default-directory (_checker)
+  "Come up with a suitable default directory to run CHECKER in.
```
This could say "to run Dogma in", since `_checker` is in fact ignored.

kthelgason replied:
As with the names, I'd prefer to keep the functions that can be reused checker-agnostic.

cpitclaudel commented:
This looks very good. I commented on the diff about a few typos.
Question: why are the functions called `flycheck-elixir` rather than `flycheck-dogma`? (this could be a convention that I'm just not aware of :)

kthelgason commented:
@cpitclaudel I'll try to address your comments later today or tomorrow, thanks for looking over the PR.
The reason I went with that naming (inspired by some other checker, can't recall which now) was that I hope to add support for more elixir checkers (credo) soon, in which case functions such as `flycheck-elixir--find-default-directory` could be shared.

lunaryorn commented:
I don't mind the name; these functions looks sufficiently generic imho to be named after the language.

cpitclaudel commented:
Thanks for the clarification (and for your efforts!). Everything LGTM modulo the small typos :)

kthelgason commented:
I fixed the typos and tried to come up with excuses for the comments I ignored 😄
Please give this one last look, Thanks!

cpitclaudel commented:
Looks perfect. @lunaryorn?

lunaryorn commented:
Uh, sh**, I knew I had forgotten a pull request for the last release :cry:  

LGTM, but unfortunately `CHANGES.rst` is now out of date as we're now heading towards Flycheck 30. 😞 

@kthelgason I'm so sorry to bother you again but could you just update `CHANGES.rst` to note the change under Flycheck 30 before you merge? I'm sorry for all the nuisance 😞

kthelgason commented:
Don't worry about it! I fixed `CHANGES.rst` (I think) and pushed. 😄

lunaryorn commented:
Feel free to merge then 😊
